File: C_files/Lecture_1/credit/credit.c
#include <stdio.h>
#include <stdbool.h>

//Function Declaration
bool checksum (long long input);
int sum_of_digits(int num);
int number_of_digits(long long input);
bool american_express(long long input);
bool mastercard(long long input);
bool visa(long long input);

int main()
{
    //Get input from user
    long long input;
    printf("Number: ");
    scanf("%lld", &input);

    //Check conditions
    if (checksum(input))
    {
        //Check American Express
        if (american_express(input))
        {
            printf("AMEX\n");
        }
        
        //Check Mastercard
        else if (mastercard(input))
        {
            printf("MASTERCARD\n");
        }

        //Check VISA
        else if (visa(input))
        {
            printf("VISA\n");
        }

        //If checksum is valid but it is none of the cards
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

//Luhn's Algorithm function
bool checksum (long long input)
{
    int final_sum = 0;

    //Calculate sum of digits of even place digit multiply by 2
    long long temp1 = input;
    while (temp1 != 0)
    {
        temp1 = temp1 / 10;
        int digit = temp1 % 10;
        int mul = (digit * 2);
        final_sum += sum_of_digits(mul);
        temp1 = temp1 / 10;
    }

    //Calculate sum of odd placed digits
    long long temp2 = input;
    while (temp2 != 0)
    {
        final_sum += temp2 % 10;
        temp2 = temp2 / 100;
    }

    //Return true of false based on Luhn's algorithm
    if (final_sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

//Sum of digits function
int sum_of_digits(int num)
{
    int sum = 0;
    long long temp = num;
    while (temp != 0)
    {
        sum += (temp % 10);
        temp = temp / 10;
    }
    return sum;
}

//Number of digits function
int number_of_digits(long long input)
{
    int digits = 0;
    while (input != 0)
    {
        input = input / 10;
        digits++;
    }
    return digits;
}

//American Express Function
bool american_express(long long input)
{
    if (number_of_digits(input) == 15)
    {
        if (input / 10000000000000 == 34 || input / 10000000000000 == 37)
        {
            return true;
        }
        return false;
    }
    return false;
}

//Mastercard Function
bool mastercard(long long input)
{
    if (number_of_digits(input) == 16)
    {
        if (input / 100000000000000 == 51 || input / 100000000000000 == 52 || input / 100000000000000 == 53 || input / 100000000000000 == 54 || input / 100000000000000 == 55)
        {
            return true;
        }
        return false;
    }
    return false;
}

//VISA Function
bool visa(long long input)
{
    if (number_of_digits(input) == 13)
    {
        if (input / 1000000000000 == 4 || input / 1000000000000 == 4)
        {
            return true;
        }
        return false;
    }
    else if (number_of_digits(input) == 16)
    {
        if (input / 1000000000000000 == 4 || input / 1000000000000000 == 4)
        {
            return true;
        }
        return false;
    }
    return false;
}

File: Python_files/Lecture_6/sentimental-credit/credit.py
def checksum(input):
    final_sum = 0

    # Calculate sum of digits of even-placed digits multiplied by 2
    temp1 = input
    while temp1 != 0:
        temp1, digit = divmod(temp1, 10)
        mul = digit * 2
        final_sum += sum_of_digits(mul)

    # Calculate sum of odd-placed digits
    temp2 = input
    while temp2 != 0:
        final_sum += temp2 % 10
        temp2 //= 100

    # Return True or False based on Luhn's algorithm
    return final_sum % 10 == 0


def sum_of_digits(num):
    return sum(int(digit) for digit in str(num))


def number_of_digits(input):
    return len(str(input))


def american_express(input):
    if number_of_digits(input) == 15:
        return input // 10000000000000 in (34, 37)
    return False


def mastercard(input):
    if number_of_digits(input) == 16:
        return input // 100000000000000 in (51, 52, 53, 54, 55)
    return False


def visa(input):
    if number_of_digits(input) == 13:
        return input // 1000000000000 == 4
    elif number_of_digits(input) == 16:
        return input // 1000000000000000 == 4
    return False


def main():
    # Get input from the user
    input_value = int(input("Number: "))

    # Check conditions
    if checksum(input_value):
        # Check American Express
        if american_express(input_value):
            print("AMEX")
        # Check Mastercard
        elif mastercard(input_value):
            print("MASTERCARD")
        # Check VISA
        elif visa(input_value):
            print("VISA")
        # If checksum is valid but it is none of the cards
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
